File: code/4958/4958接龙数列.py
if __name__ == "__main__":
    n = int(input())
    nums = list(input().split())
    g = [0] * 10
    res = 0
    for i in range(n):
        num = nums[i]
        l, r = ord(num[0]) - ord('0'), ord(num[-1]) - ord('0')
        f = max(1, g[l] + 1)
        g[r] = max(g[r], f)
        res = max(res, f)
    print(n - res)

File: code/4958/4958接龙数列.cpp
#include <iostream>
#include <cstring>
#include <algorithm>
using namespace std;

const int N = 1e5 + 10;
int g[N];
int n, res;
string str;

int main() {
	cin >> n;

	for (int i = 0; i < n; i++) {
		cin >> str;
		int l = str[0] - '0', r = str[str.size() - 1] - '0';
		int f = max(1, g[l] + 1);
		g[r] = max(g[r], f);
		res = max(res, f);
	}

	cout << n - res << endl;

	return 0;
}
